### PR TITLE
Functional algorithms for async generators

### PR DESCRIPTION
**Pull request by bobvanmanen**
Provides a set of functional algorithms to be used on async generators. 
These provide the basic building blocks to allow us to concurrently upload and download (parts of) files.

### CODE REVIEW

File: lib/headers.js [hunk 1]
```diff
@@ -27,49 +28,155 @@ async function getHeaders(url, options={}) {
...
+ * Retrieve and parse Content-Range from the HTTP headers
+ * 
+ * @param {Headers} headers HTTP headers
+ * @returns {ContentRange} Content range
+ */
+function getContentRange(headers) {
+    const contentRange = headers.get(HTTP.HEADER.CONTENT_RANGE);
+    if (contentRange) {
+        return parseContentRange(contentRange);
+    }
```
Comment (mfrisbey) [inline lib/headers.js:151]: Missing return value if condition not met?

File: lib/headers.js [hunk 2]
```diff
@@ -27,49 +28,155 @@ async function getHeaders(url, options={}) {
...
+function getContentLength(headers) {
+    const value = headers.get(HTTP.HEADER.CONTENT_LENGTH);
+    if (value) {
+        const parsed = parseInt(value);
+        if (typeof parsed === "number" && !isNaN(parsed)) {
+            return parsed;
         }
-    } else {
-        return null;
     }
```
Comment (mfrisbey) [inline lib/headers.js:167]: Missing return value if condition not met?

**Review by mfrisbey — CHANGES_REQUESTED** [review]
Looks great, just a couple potential missing return values.

**Comment by bobvanmanen** [comment]
> Looks great, just a couple potential missing return values.

They're not strictly necessary, but without the return values it is a bit confusing. These changes were part of an earlier PR. I've opened a PR to fix it though: https://github.com/adobe/node-httptransfer/pull/41

**Review by mfrisbey — APPROVED** [review]
Ugh, javascript :P

Agreed, they're not necessary, but also agreed that it can be confusing.

File: lib/generator/function.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ * governing permissions and limitations under the License.
+ */
+
+"use strict";
+
+/**
+ * Base async generator function class that can be executed on a set of items,
+ * which yields the same number of results as input items.
+ */
+class AsyncGeneratorFunction {
```
Comment (tmathern) [inline lib/generator/function.js:19]: I think `AsyncGenerator` would have been enough. I think only functions can be generators anyway...

File: lib/generator/mapconcurrent.js [hunk 4]
```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2020 Adobe. All rights reserved.
```
Comment (tmathern) [inline lib/generator/mapconcurrent.js:2]: 2021

File: lib/generator/pipeline.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2020 Adobe. All rights reserved.
```
Comment (tmathern) [inline lib/generator/pipeline.js:2]: 2021

File: test/generator.test.js [hunk 6]
```diff
@@ -0,0 +1,526 @@
+/*
+ * Copyright 2020 Adobe. All rights reserved.
```
Comment (tmathern) [inline test/generator.test.js:2]: 2021

File: test/generator.test.js [hunk 7]
```diff
@@ -0,0 +1,526 @@
...
+                const result = await executeBatch({
+                    execute: async function*(values) {
+                        for await (const value of values) {
+                            yield value + 1;
+                        }
+                    }
+                }, [ ]);
+                assert.deepStrictEqual(result, [ ]);
+            });
+            it('array', async function() {
```
Comment (tmathern) [inline test/generator.test.js:119]: In the code, I didn't see checks that arrays are used. What if it gets Sets or Maps?

File: test/generator.test.js [hunk 8]
```diff
@@ -0,0 +1,526 @@
...
+                    execute: async function*(values) {
+                        for await (const value of values) {
+                            yield value + 1;
+                        }
+                    }
+                }, toGenerator([ 1 ]));
+                assert.deepStrictEqual(result, [ 2 ]);
+            });
+        });
+        describe('waitResult', function() {
```
Comment (tmathern) [inline test/generator.test.js:150]: I would add a test in which `pending` is empty.

**Review by tmathern — COMMENTED** [review]
LGTM, but check comments.

File: lib/generator/mapconcurrent.js [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+        // Ensure the original pending promise is returned with the value, so we know
+        // which promise to delete from the pending array.
+        const { promise, result } = await Promise.race(pending.map(promise =>
+            Promise.resolve(promise).then(result => {
+                return { promise, result };
+            })
+        ));
+
+        const i = pending.findIndex(pendingPromise => pendingPromise === promise);
+        assert.ok(i >= 0, `Promise ${promise} -> ${result} not in pending state`);
```
[the remark this answers is not in the document]
Comment (bobvanmanen) [reply lib/generator/mapconcurrent.js:40]: No, it's intended to be there and detect a failure in the promise handling. It used to have an if statement and a `throw Error` but there is no way for me to build a unit test for it so `codecov` failed my PR.

File: test/generator.test.js (hunk 7 above)
Comment (bobvanmanen) [reply test/generator.test.js:119]: That should be fine, as long as `for await (const v of items) { }` works which is true for Set or Map.
